sporadic compilation failures with parallel builds enabled

For time to time, building `bsc` with a huge number of jobs `-j48` fails.

I managed to get a log output for one of these failures:
[bluespec-bsc-parallel-build-failure.txt](https://github.com/B-Lang-org/bsc/files/4175175/bluespec-bsc-parallel-build-failure.txt)

The interesting lines:

```
[  4 of 149] Compiling EitherUtil       ( EitherUtil.hs, /build/source/src/comp/../../build/EitherUtil.o )
[  4 of 149] Compiling EitherUtil       ( EitherUtil.hs, /build/source/src/comp/../../build/EitherUtil.o )
[  5 of 149] Compiling ErrorTCompat     ( ErrorTCompat.hs, /build/source/src/comp/../../build/ErrorTCompat.o )
[  5 of 149] Compiling ErrorTCompat     ( ErrorTCompat.hs, /build/source/src/comp/../../build/ErrorTCompat.o )
[  1 of 224] Compiling BDD              ( BDD.hs, /build/source/src/comp/../../build/BDD.o )
/build/source/src/comp/../../build/ErrorTCompat.o: getModificationTime:getFileStatus: does not exist (No such file or directory)
[  6 of 149] Compiling Eval             ( Eval.hs, /build/source/src/comp/../../build/Eval.o )
make[2]: *** [Makefile:403: bluetcl] Error 1
make[2]: *** Waiting for unfinished jobs....
```

It seems there's at least two builds ghc builds going on in parallel (one with 149 targets, and one with 224 targets), and they both manage to stumble over each others feets trying to read/write to/from `ErrorTCompat`.

We should probably run only one of these builds at once, have intermediate build artifacts in separate scratch dirs, or further split things into libraries (as suggested by @NinjaTrappeur in https://github.com/B-Lang-org/bsc/issues/22)

--- a/src/Makefile
+++ b/src/Makefile
@@ -11,7 +11,8 @@ install clean full_clean:
 	$(MAKE)  -C stp        PREFIX=$(PREFIX)  $@
 	$(MAKE)  -C yices      PREFIX=$(PREFIX)  $@
 	$(MAKE)  -C tcltk      PREFIX=$(PREFIX)  $@
-	$(MAKE)  -C comp       PREFIX=$(PREFIX)  $@
+	# we need to build targets from here sequentially, as they operate in the same workspace
+	$(MAKE)  -C comp -j1   PREFIX=$(PREFIX)  $@
 	$(MAKE)  -C Libraries  PREFIX=$(PREFIX)  $@
 	$(MAKE)  -C exec       PREFIX=$(PREFIX)  $@
 	$(MAKE)  -C VPI        PREFIX=$(PREFIX)  $@

--- a/src/comp/Makefile
+++ b/src/comp/Makefile
@@ -79,6 +79,7 @@ WISHFLAGS += $(EXTRAWISHLIBS)
 # GHC
 
 GHC ?= ghc
+GHCJOBS ?= 1
 
 GHCVERSION=$(shell $(GHC) --version | head -1 | egrep -o "[0-9]+\.[0-9]+\.[0-9]" | head -1)
 
@@ -193,7 +194,7 @@ PACKAGES = \
 # The make flags do not include "-o" because bluetcl and bluewish
 # are compiled in two steps and the first step doesn't have that option.
 # So all users of GHCMAKEFLAGS have to include "-o" themselves.
-GHCMAKEFLAGS = --make $@
+GHCMAKEFLAGS = --make $@ -j$(GHCJOBS)
 GHCCOMPILEFLAGS = -o $@ -c $<
 
 # On Mac OS, we'll need to update the dylib location,

